### ISSUE TITLE
Runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized

### ISSUE THREAD

**Issue by mattmattox** [rancher/rancher#23254]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
- Deploy Rancher v2.2.8
- Create custom cluster with default settings
- Add node with missing firewall rules so node will only been partially added
- remove node from Rancher UI and try to rejoin

**Result:**
Runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialize

**Other details that may be helpful:**
This looks very similar to https://github.com/rancher/rancher/issues/13484

**Workaround**
Copy /etc/cni/net.d/10-canal.conflist and calico-kubeconfig from working node to new node. Then join node to cluster.

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): v2.2.8
- Installation option (single install/HA): HA

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Custom
- Machine type (cloud/VM/metal) and specifications (CPU/memory): 2/4
- Kubernetes version (use `kubectl version`):

```
v1.14.6-rancher1
```

- Docker version (use `docker version`):

```
17.03.2-ce
```

**Comment by alena1108** [comment]
For RKE clusters provisioned with Rancher, there is a special handling for worker nodes. Node update/delete is handled by rancher agent, and is not included in the main rke reconcile loop. For the ref, that's where this exclusion happens https://github.com/rancher/rancher/blob/master/pkg/controllers/management/clusterprovisioner/driver.go#L165.

On the node creation, rancher agent gets a node plan from rke/rancher, places the configuration on the node and start all the k8s components https://github.com/rancher/rancher/blob/master/pkg/controllers/user/noderemove/nodes.go#L30
On node removal, agent simply removes the v1.node, and then rancher kills agent container https://github.com/rancher/rancher/blob/master/pkg/controllers/management/node/controller.go#L255.

We have to add a cleanup part to the node removal logic on Rancher/agent side, and remove all the necessary configs before proceeding with the agent removal.

Keep in mind that the cleanup will be performed only when node can be contacted. If the node is unavailable at the moment of agent removal call, manual cleanup will be needed before the node could be re-registered with Rancher. As per https://rancher.com/docs/rancher/v2.x/en/cluster-admin/cleaning-cluster-nodes/.

**Comment by yhyddr** [comment]
For me, just some hours later and everything would be good.

**Comment by YuanChenLi** [comment]
check your hostname

**Comment by fengzixu** [comment]
> For RKE clusters provisioned with Rancher, there is a special handling for worker nodes. Node update/delete is handled by rancher agent, and is not included in the main rke reconcile loop. For the ref, that's where this exclusion happens https://github.com/rancher/rancher/blob/master/pkg/controllers/management/clusterprovisioner/driver.go#L165.
> 
> On the node creation, rancher agent gets a node plan from rke/rancher, places the configuration on the node and start all the k8s components https://github.com/rancher/rancher/blob/master/pkg/controllers/user/noderemove/nodes.go#L30
> 
> On node removal, agent simply removes the v1.node, and then rancher kills agent container https://github.com/rancher/rancher/blob/master/pkg/controllers/management/node/controller.go#L255.
> 
> We have to add a cleanup part to the node removal logic on Rancher/agent side, and remove all the necessary configs before proceeding with the agent removal.
> 
> Keep in mind that the cleanup will be performed only when node can be contacted. If the node is unavailable at the moment of agent removal call, manual cleanup will be needed before the node could be re-registered with Rancher. As per https://rancher.com/docs/rancher/v2.x/en/cluster-admin/cleaning-cluster-nodes/.

Can you tell us which config files need to be cleaned for this issue ? @alena1108

**Comment by PaoloneM** [comment]
Same problem while trying to restore the master node in a Disaster Recovery scenario. 

Steps to reproduce:

- Create cluster with RKE 1.0.4
- Wipe master node with superseb's extended cleanup script
- Re-run `rke up` to restore master node

Master node comes up with the error in issue title, then, after restoring /etc/cni/net.d/10-canal.conflist and calico-kubeconfig from backup or working node I got the following error in kubelet:

```
W0211 10:01:54.192001   17287 cni.go:202] Error validating CNI config &{k8s-pod-network 0.3.1 false [0xc001471b00 0xc001471b80] [123 10 32 32 34 110 97 109 101 34 58 32 34 107 56 115 45 112 111 100 45 110 101 116 119 111 114 107 34 44 10 32 32 34 99 110 105 86 101 114 115 105 111 110 34 58 32 34 48 46 51 46 49 34 44 10 32 32 34 112 108 117 103 105 110 115 34 58 32 91 10 32 32 32 32 123 10 32 32 32 32 32 32 34 116 121 112 101 34 58 32 34 99 97 108 105 99 111 34 44 10 32 32 32 32 32 32 34 108 111 103 95 108 101 118 101 108 34 58 32 34 87 65 82 78 73 78 71 34 44 10 32 32 32 32 32 32 34 100 97 116 97 115 116 111 114 101 95 116 121 112 101 34 58 32 34 107 117 98 101 114 110 101 116 101 115 34 44 10 32 32 32 32 32 32 34 110 111 100 101 110 97 109 101 34 58 32 34 109 97 115 116 101 114 34 44 10 32 32 32 32 32 32 34 105 112 97 109 34 58 32 123 10 32 32 32 32 32 32 32 32 32 32 34 116 121 112 101 34 58 32 34 104 111 115 116 45 108 111 99 97 108 34 44 10 32 32 32 32 32 32 32 32 32 32 34 115 117 98 110 101 116 34 58 32 34 117 115 101 80 111 100 67 105 100 114 34 10 32 32 32 32 32 32 125 44 10 32 32 32 32 32 32 34 112 111 108 105 99 121 34 58 32 123 10 32 32 32 32 32 32 32 32 32 32 34 116 121 112 101 34 58 32 34 107 56 115 34 44 10 32 32 32 32 32 32 32 32 32 32 34 107 56 115 95 97 117 116 104 95 116 111 107 101 110 34 58 32 34 101 121 74 104 98 71 99 105 79 105 74 83 85 122 73 49 78 105 73 115 73 109 116 112 90 67 73 54 73 109 90 77 85 71 53 74 78 110 107 121 84 105 49 72 86 68 73 122 90 49 82 49 82 106 74 80 86 107 90 121 87 87 57 51 89 88 74 66 84 50 107 120 88 50 82 117 97 108 78 48 90 84 99 53 86 88 77 105 102 81 46 101 121 74 112 99 51 77 105 79 105 74 114 100 87 74 108 99 109 53 108 100 71 86 122 76 51 78 108 99 110 90 112 89 50 86 104 89 50 78 118 100 87 53 48 73 105 119 105 97 51 86 105 90 88 74 117 90 88 82 108 99 121 53 112 98 121 57 122 90 88 74 50 97 87 78 108 89 87 78 106 98 51 86 117 100 67 57 117 89 87 49 108 99 51 66 104 89 50 85 105 79 105 74 114 100 87 74 108 76 88 78 53 99 51 82 108 98 83 73 115 73 109 116 49 89 109 86 121 98 109 86 48 90 88 77 117 97 87 56 118 99 50 86 121 100 109 108 106 90 87 70 106 89 50 57 49 98 110 81 118 99 50 86 106 99 109 86 48 76 109 53 104 98 87 85 105 79 105 74 106 89 87 53 104 98 67 49 48 98 50 116 108 98 105 49 52 99 71 82 119 98 67 73 115 73 109 116 49 89 109 86 121 98 109 86 48 90 88 77 117 97 87 56 118 99 50 86 121 100 109 108 106 90 87 70 106 89 50 57 49 98 110 81 118 99 50 86 121 100 109 108 106 90 83 49 104 89 50 78 118 100 87 53 48 76 109 53 104 98 87 85 105 79 105 74 106 89 87 53 104 98 67 73 115 73 109 116 49 89 109 86 121 98 109 86 48 90 88 77 117 97 87 56 118 99 50 86 121 100 109 108 106 90 87 70 106 89 50 57 49 98 110 81 118 99 50 86 121 100 109 108 106 90 83 49 104 89 50 78 118 100 87 53 48 76 110 86 112 90 67 73 54 73 106 65 120 89 84 74 109 78 50 90 107 76 84 66 104 90 68 103 116 78 71 82 104 90 83 48 53 89 109 77 52 76 84 77 121 78 68 78 107 89 106 82 105 79 84 85 119 90 67 73 115 73 110 78 49 89 105 73 54 73 110 78 53 99 51 82 108 98 84 112 122 90 88 74 50 97 87 78 108 89 87 78 106 98 51 86 117 100 68 112 114 100 87 74 108 76 88 78 53 99 51 82 108 98 84 112 106 89 87 53 104 98 67 74 57 46 107 66 114 95 78 109 107 102 71 70 66 72 88 78 111 65 69 89 57 119 85 97 48 98 103 83 70 106 49 71 101 71 80 116 48 72 114 48 104 112 121 107 98 90 76 106 66 53 57 77 51 53 51 106 48 106 112 75 65 49 109 77 55 77 112 74 114 50 90 104 88 120 119 49 68 50 122 117 116 121 117 118 67 48 121 78 107 119 100 75 118 111 95 49 50 98 75 98 81 95 54 116 54 104 66 67 99 48 97 97 90 57 69 48 104 90 80 51 99 119 54 78 115 110 45 71 90 85 55 110 108 80 69 80 72 54 82 80 108 45 85 117 97 105 122 110 95 112 86 87 82 54 74 74 67 72 65 117 89 121 122 69 76 109 71 102 65 95 122 76 95 116 69 72 50 98 57 102 53 99 81 121 56 85 75 49 105 52 51 82 85 78 65 83 72 65 99 65 56 65 57 105 109 108 118 88 108 89 119 56 101 78 68 88 83 85 110 105 86 49 49 67 68 98 51 108 103 69 57 45 66 72 89 109 120 113 116 52 68 81 122 83 83 116 103 122 86 45 88 88 101 70 55 120 73 77 82 89 65 105 106 65 72 80 67 79 106 119 120 117 82 56 56 98 51 53 106 121 114 51 45 74 108 72 109 57 95 51 49 107 95 57 100 114 100 113 52 76 85 95 88 116 105 120 85 54 49 86 87 99 68 56 76 49 72 50 70 73 45 78 90 49 99 111 118 90 51 48 99 82 71 122 108 122 111 110 74 122 104 69 71 119 34 10 32 32 32 32 32 32 125 44 10 32 32 32 32 32 32 34 107 117 98 101 114 110 101 116 101 115 34 58 32 123 10 32 32 32 32 32 32 32 32 34 107 117 98 101 99 111 110 102 105 103 34 58 32 34 47 101 116 99 47 107 117 98 101 114 110 101 116 101 115 47 115 115 108 47 107 117 98 101 99 102 103 45 107 117 98 101 45 110 111 100 101 46 121 97 109 108 34 10 32 32 32 32 32 32 125 10 32 32 32 32 125 44 10 32 32 32 32 123 10 32 32 32 32 32 32 34 116 121 112 101 34 58 32 34 112 111 114 116 109 97 112 34 44 10 32 32 32 32 32 32 34 115 110 97 116 34 58 32 116 114 117 101 44 10 32 32 32 32 32 32 34 99 97 112 97 98 105 108 105 116 105 101 115 34 58 32 123 34 112 111 114 116 77 97 112 112 105 110 103 115 34 58 32 116 114 117 101 125 10 32 32 32 32 125 10 32 32 93 10 125 10]}: [failed to find plugin "calico" in path [/opt/cni/bin] failed to find plugin "portmap" in path [/opt/cni/bin]]
W0211 10:01:54.192410   17287 cni.go:237] Unable to update cni config: no valid networks found in /etc/cni/net.d
E0211 10:01:55.670086   17287 kubelet.go:2187] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
```

**Comment by StrongMonkey** [comment]
I think there is a general race condition on this issue. 

RKE relies on deploying networkPlugin add-ons(listed in here https://github.com/rancher/kontainer-driver-metadata/blob/dev-v2.4/rke/templates/calico.go#L114-L143) This is mounted using configmap and hostpath so that cni config is present on the host. 

NetworkPlugin is deployed as daemonset, where before cni config is initialised, Kubelet will try to find cni config, and if that fails over a period of time then kubelet will mark itself as notReady. Once kubelet is marked as unready before daemonset pod is scheduled, cni config could never be initialized into the host.

**Comment by jiaqiluo** [comment]
The bug is reproduced in rancher:v2.3.5

Steps:
- add a custom cluster with 1 etcd+contrlPlane, 2 workers
- after the cluster is active, removed worker-1 from the UI
- after the cluster is active again, add worker-1 back by running the docker command on the node

results:
- worker-1 is stuck in registering state 
<img width="1389" alt="Screen Shot 2020-03-11 at 11 27 39 AM" src="https://user-images.githubusercontent.com/6218999/76452544-8b316f00-638e-11ea-9b82-6bff3b341bbc.png">

**Comment by jiaqiluo** [comment]
the bug fix is validated in rancher:master-head c34bdc2bb

Steps:
- add a custom cluster with 1 etcd+contrlPlane, 2 workers
- after the cluster is active, removed worker-1 from the UI
- after the cluster is active again, add worker-1 back by running the docker command on the node

results:
- worker-1 is added back successfully 
- the `kubelet` container in worker-1 is restarted 
- deploy a daemonSet and the pod can be assigned on worker-1 properly

**Comment by deniseschannon** [comment]
This fix was reverted in v2.4.1 cause it was causing https://github.com/rancher/rancher/issues/26399. 

We will look into a proper fix later.

**Comment by deniseschannon** [comment]
@sangeethah @jiaqiluo we need to do testing around the agent upgrades and viewing to ensure there are no issues. The original fix of this issue was what triggered the kubelet restart issue in v2.4.0 but we've put a proper fix in so that we could spend more time testing around the upgrades and trying to trigger the race condition around if the managed or unmanaged agent starts first. 

For any questions, please ask @StrongMonkey.

**Comment by jiaqiluo** [comment]
The fix needs to be forward-ported to the master branch.

**Comment by StrongMonkey** [comment]
@jiaqiluo It is already on master now https://github.com/rancher/rancher/commit/a6907ab73a595a05e5e80f84a4fe3efec440d5df.

**Comment by jiaqiluo** [comment]
Some observations on v2.4.2

First, we add a custom cluster with 1 etcd+contrlPlane, 2 workers. After the cluster is active we remove worker-2 from the UI, it turns out that:
- the output of  `kubectl get nodes`  shows that worker-2 is removed from the cluster immediately
- the container `kubelet` keeps running in the instance along with other 2 containers
```
> docker ps
CONTAINER ID        IMAGE                                COMMAND                  CREATED             STATUS              PORTS               NAMES
9db2a348f702        rancher/rke-tools:v0.1.56            "nginx-proxy CP_HOST…"   10 minutes ago      Up 10 minutes                           nginx-proxy
9d9cdbe10879        rancher/hyperkube:v1.17.4-rancher1   "/opt/rke-tools/entr…"   10 minutes ago      Up 10 minutes                           kubelet
05cea0c510aa        rancher/hyperkube:v1.17.4-rancher1   "/opt/rke-tools/entr…"   10 minutes ago      Up 10 minutes                           kube-proxy
```

- the logs of `kubelet` shows the following error:
```
E0409 17:45:19.763171   13414 controller.go:227] failed to get node "ip-172-31-26-113" when trying to set owner ref to the node lease: nodes "ip-172-31-26-113" not found
E0409 17:45:20.601269   13414 eviction_manager.go:246] eviction manager: failed to get summary stats: failed to get node info: node "ip-172-31-26-113" not found
E0409 17:45:23.872481   13414 kubelet_node_status.go:402] Error updating node status, will retry: error getting node "ip-172-31-26-113": nodes "ip-172-31-26-113" not found
```

**Comment by jiaqiluo** [comment]
The bug fix is validated on v2.4-2737-head head  8439d8bde
Steps:
- add a custom cluster with 1 etcd+contrlPlane, 2 workers
- after the cluster is active, removed worker-1 from the UI
- after the cluster is active again, add worker-1 back by running the docker command on the node

results:
- worker-1 is added back successfully 
- the `kubelet` container in worker-1 is restarted 
- deploy a daemonSet and the pod can be assigned on worker-1 properly

**Comment by jiaqiluo** [comment]
The following behavior happens on both master-head and v2.4-head:
- after removing the node from UI, all running contains are left running on the node (`kubelet` + `nginx-proxy `+ `kube-proxy` + a bunch of `k8s_*` container  )
- after running the registry command on the node, the `kubelet` container is deleted and never come back, a container using the image `rancher/rancher-agent` is started
- errors in the rancher/rancher-agent container:
```
time="2020-04-09T21:37:37Z" level=info msg="Connecting to proxy" url="wss://68.183.53.124/v3/connect"
time="2020-04-09T21:37:37Z" level=error msg="Remotedialer proxy error" error="Error response from daemon: invalid volume specification: '::rw,rslave,z'"
time="2020-04-09T21:37:47Z" level=info msg="Connecting to wss://68.183.53.124/v3/connect with token 9cxcb5r2b87z5tk2xt5kfskvx46x5fz87dvbtt2httgrcn9r65vjqx"
time="2020-04-09T21:37:47Z" level=info msg="Connecting to proxy" url="wss://68.183.53.124/v3/connect"
time="2020-04-09T21:37:47Z" level=error msg="Remotedialer proxy error" error="Error response from daemon: invalid volume specification: '::rw,rslave,z'"
```

**Comment by StrongMonkey** [comment]
@jiaqiluo I think I found the issue, but just curious what OS and docker version you are using? The logs `invalid volume specification: '::rw,rslave,z'"` does seems irrelevant to my fix. I wonder the OS or docker version is causing it.

**Comment by kinarashah** [comment]
@jiaqiluo @StrongMonkey Just saw this on @soumyalj's setup as well, where the dockerInfo for the host is temporarily empty, and the nodePlan thus passes wrong binds for the kubelet -- node never registers. We have never hit this before, @sangeethah can we look into what specific OS or docker version is causing this behavior?

**Comment by jiaqiluo** [comment]
@StrongMonkey @kinarashah 
docker version
```
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b7f0
 Built:             Wed Mar 11 01:25:46 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b7f0
  Built:            Wed Mar 11 01:24:19 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

OS info
```
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.3 LTS
Release:	18.04
Codename:	bionic
```

**Comment by jiaqiluo** [comment]
The bug fix is validated in v2.4-head 63a490f and master-head 00ff159


- after removing the node from UI, containers whose name starts with `k8s_` are stopped in the node 
- after running the resigner command in the node, a container using the image `rancher/rancher-agent` is up
```
ubuntu@ip-172-31-7-154:~$ docker ps
CONTAINER ID        IMAGE                                  COMMAND                  CREATED             STATUS                                  PORTS               NAMES
4cea9c17ce3c        rancher/rancher-agent:v2.4-2773-head   "run.sh --server htt…"   2 seconds ago       Restarting (1) Less than a second ago                       youthful_ride
```

- this above-mentioned container restarts the container named `kubelet`, then it is stoped and deleted ( it is  gone in the output of  `docker ps -a`)
- note that the `kubelet` is still the existing container, it gets restarted only one time

```
cf7a9bf82ed3        rancher/hyperkube:v1.17.4-rancher1     "/opt/rke-tools/entr…"   36 minutes ago      Up 22 minutes                                   kubelet
```
